### ISSUE TITLE
Rethink `UIDArray.__getitem__()`

### ISSUE THREAD

**Issue by cliffckerr** [starsimhub/starsim#423]
Currently, it takes the vast majority of time in `make_new_cases()`:
![image](https://github.com/starsimhub/starsim/assets/3239256/c7d801cc-03df-47a2-b5c5-314b0a207ae4)
(See https://github.com/starsimhub/starsim/issues/422 for more performance tests.)

Is there a way we could use more standard Numpy operations to make this faster, or reduce some of the checking? It's already Numbafied, so can't get any additional speed increases there.

Another option would be to go back to not removing agents, so the UID is also just the array index.

**Comment by cliffckerr** [comment]
Another benchmarking:
```py
import sciris as sc
import starsim as ss

with sc.cprofile(sort='selfpct') as cpr:
    sim = ss.demo(n_agents=100e3, start=2000, end=2100, diseases='sis', plot=False)

sim.diseases.sis.plot()
```
gives
```
                                                  func    cumpct     selfpct  cumtime    selftime  calls                                         path
0                                       _get_vals_uids   35.5618  3.5518e+01   1.5462  1.5443e+00   1317                                 states.py:83
1    <method 'permutation' of 'numpy.random._genera...   12.5981  1.2597e+01   0.5478  5.4770e-01    102                                            ~
2                                       make_new_cases   61.5934  8.5633e+00   2.6780  3.7233e-01    101                               disease.py:255
3                                      __array_ufunc__    8.2585  8.1880e+00   0.3591  3.5601e-01   3950                                states.py:267
4                                                  rvs   27.3114  4.3405e+00   1.1875  1.8872e-01    509                         distributions.py:440
5                                            add_pairs   19.5998  4.1161e+00   0.8522  1.7897e-01    102                               network.py:497
6    <method 'lognormal' of 'numpy.random._generato...    3.0262  3.0262e+00   0.1316  1.3158e-01    102                                            ~
7                                          beta_per_dt    2.3906  2.3694e+00   0.1039  1.0302e-01    202                               network.py:288
8    <method 'random' of 'numpy.random._generator.G...    2.0724  2.0724e+00   0.0901  9.0107e-02    406                                            ~
9           <method 'reduce' of 'numpy.ufunc' objects>    1.6476  1.6434e+00   0.0716  7.1452e-02    821                                            ~
10                                           end_pairs    6.6629  1.1970e+00   0.2897  5.2043e-02    101                               network.py:324
```

**Comment by cliffckerr** [comment]
No easy fixes here, pure NumPy is slightly slower than Numba. But considering that 35% of the entire sim time is spent on this function, I wonder if we would be better off not removing agents and just going back to array indexing with UID == position in array.

**Comment by cliffckerr** [comment]
Frustratingly, this parallelized version is 3x faster for large sims _but_ 2x slower for small ones!
```py
    @staticmethod
    @nb.njit(cache=True, parallel=False)
    def _get_vals_uids(vals, key, uid_map):
        """
        Extract values from a collection of UIDs

        This function is used to retrieve values based on UID. As indexing a UIDArray returns a new UIDArray,
        this method also populates the new UID map for use in the subsequently created UIDArray, avoiding the
        need to re-compute it separately.

        :param vals: A 1D np.ndarray containing the values
        :param key: A 1D np.ndnarray of integers containing the UIDs to query
        :param uid_map: A 1D np.ndarray of integers mapping UID to array position in ``vals``
        :return: A tuple of (values, uids, new_uid_map) suitable for passing into the UIDArray constructor
        """
        len_key = len(key)
        out = np.empty(len_key, dtype=vals.dtype)
        new_uid_map = np.full(uid_map.shape[0], fill_value=INT_NAN, dtype=np.int64)

        for i in nb.prange(len_key):
            kv = key[i]
            idx = uid_map[kv]
            out[i] = vals[idx]
            new_uid_map[kv] = i
        return out, key, new_uid_map
```

**Comment by cliffckerr** [comment]
Added a switch between the two ... not perfect but seems to work!

**Comment by cliffckerr** [comment]
Readded priority label given that the current approach doesn't work with deaths; see #296

**Comment by cliffckerr** [comment]
Explored NumPy `masked_array`, but insanely inefficient:
```py
import numpy as np
import sciris as sc

reps = 100
n = int(1e7)
m = int(1e5)

a = np.random.rand(n)
inds = np.random.randint(n, size=m)

with sc.timer('indexing'):
    for r in range(reps):
        a[inds].sum()


mask = np.ones(n)
mask[inds] = False
b = np.ma.masked_array(a, mask)

with sc.timer('masked'):
    for r in range(reps):
        b.compressed().sum()

indexing: 16.8 ms
masked: 0.366 s
```

**Comment by cliffckerr** [comment]
One option:
```py
import numpy as np
import sciris as sc
np.random.seed(1)


class LittlePeople:
    def __init__(self, n):
        self.n = n
        self.uid = np.arange(n)
        self.age = np.random.uniform(0, 100, n)
        self.sex = np.random.rand(n) < 0.5
        self.inds = np.arange(n)
        
    def __getattribute__(self, key):
        if key in ['age', 'sex']:
            return super().__getattribute__(key)[self.inds]
        else:
            return super().__getattribute__(key)
    
    def get(self, key, inds):
        return super().__getattribute__(key)[inds]
        
    def reset(self):
        self.inds = np.arange(self.n)
        

class View:
    def __init__(self, ppl):
        self.ppl = ppl
        self.inds = sc.dcp(self.ppl.inds)
    
    def __getattribute__(self, key):
        if key in ['age', 'sex', 'uid']:
            return self.ppl.get(key, self.inds)
        else:
            return super().__getattribute__(key)
    
    def filter(self, cond, verbose=True):
        if verbose:
            print(f'Running filter: len={len(cond)}, true={sum(cond)}')
        self.inds = self.inds[cond]
        return self
    
    def __call__(self, cond):
        return self.filter(cond)
        

ppl = LittlePeople(100)
ppl.inds = np.arange(50)
vi = View(ppl)
print(vi.age)

max_age = 20
sex = 1
uids = vi(vi.age < max_age)(vi.sex==sex).uid

assert all(ppl.age[uids] < max_age)
assert all(ppl.sex[uids] == sex)
```

**Comment by RomeshA** [comment]
On my machine, running `sim = ss.demo(n_agents=100e3, start=2000, end=2100, diseases='sis', plot=False)` the parallel version takes around 7.4s and the non-parallel version takes 6.7s. So it might not be generally beneficial to have that parallel switching? On my other computer it's 3.6s non-parallel and 3.4s parallel so there is a small benefit but seems only marginal (and presumably this would be lost if running sims in parallel anyway?)

It would be important though to also test the use cases that we originally envisaged for removing dead agents since the performance tradeoff might be different for those

**Comment by cliffckerr** [comment]
Closed by #456
